src/ripple/basics/feeunits.h

```diff
@@ -0,0 +1,540 @@
+
+    template<class Other>
+    using enable_if_compatible_t =
+        typename std::enable_if_t<is_compatible_v<Other>>;
+
+    template<class OtherFee>
+    using enable_if_compatiblefee_t =
+        typename std::enable_if_t<is_compatiblefee_v<OtherFee>>;
+
+public:
+    TaggedFee () = default;
+    constexpr TaggedFee (TaggedFee const& other) = default;
+    constexpr TaggedFee& operator= (TaggedFee const& other) = default;
+
+    // Little trick stolen from boost::units that allows a 0 in the
+    // implicit constructor, but no other value
+    constexpr TaggedFee(unspecified_null_pointer_constant_type) : fee_()
+    {
+    }
+
+    constexpr
+    explicit
+    TaggedFee (beast::Zero)
+        : fee_ (0)
+    {
+    }
+
+    constexpr
+    TaggedFee&
+    operator= (beast::Zero)
+    {
+        fee_ = 0;
+        return *this;
+    }
+
+    constexpr
+    explicit
+    TaggedFee (value_type fee)
+        : fee_ (fee)
+    {
+    }
+
+    template <class Other,
+        class = enable_if_compatible_t <Other>>
+    constexpr
+    explicit
+    TaggedFee (Other fee)
+        : fee_ (static_cast<value_type> (fee))
+    {
+    }
+
+    template <class OtherFee,
+        class = enable_if_compatiblefee_t<OtherFee> >
+    constexpr
+    TaggedFee (OtherFee const& fee)
+        : fee_ (static_cast<value_type> (fee.value()))
+    {
+    }
+
+    template <class Other,
+        class = enable_if_compatible_t <Other>>
+    TaggedFee&
+    operator= (Other fee)
+    {
+        fee_ = static_cast<value_type> (fee);
+        return *this;
+    }
+
+    template <class OtherFee,
+        class = enable_if_compatiblefee_t <OtherFee>>
+    TaggedFee&
+    operator= (OtherFee const& fee)
+    {
+        fee_ = static_cast<value_type> (fee.value());
+        return *this;
+    }
+
+    constexpr
+    TaggedFee
+    operator* (value_type const& rhs) const
+    {
+        return TaggedFee{ fee_ * rhs };
+    }
+
+    constexpr
+    value_type
+    operator/ (TaggedFee const& rhs) const
+    {
+        return fee_ / rhs.fee_;
+    }
+
+    TaggedFee&
+    operator+= (TaggedFee const& other)
+    {
+        fee_ += other.fee();
+        return *this;
+    }
+
+    TaggedFee&
+    operator-= (TaggedFee const& other)
+    {
+        fee_ -= other.fee();
+        return *this;
+    }
+
+    TaggedFee&
+    operator++()
+    {
+        ++fee_;
+        return *this;
+    }
+
+    TaggedFee&
+    operator--()
+    {
+        --fee_;
+        return *this;
+    }
+
+    TaggedFee&
+    operator*= (value_type const& rhs)
+    {
+        fee_ *= rhs;
+        return *this;
+    }
+
+    TaggedFee&
+    operator/= (value_type const& rhs)
+    {
+        fee_ /= rhs;
+        return *this;
+    }
+
+    TaggedFee&
+    operator%= (value_type const& rhs)
+    {
+        fee_ %= rhs;
+        return *this;
+    }
+
+    TaggedFee
+    operator- () const
+    {
+        static_assert( std::is_unsigned_v<T>,
```

Comment:
Should this be `is_signed_v<value_type>` -> assert that it is signed?

Reply:
WAT!  Nice catch! Fixed, and I'll add unit tests.